File: VB/HeadFirstPatterns/DoFactory.HeadFirst/DoFactory.HeadFirst.Factory.Abstract.Pizza/PizzaTestDrive.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Text

Namespace DoFactory.HeadFirst.Factory.Abstract.Pizza
	' PizzaTestDrive test application
	Friend Class PizzaTestDrive
        Shared Sub Main(ByVal args() As String)

            Dim nyStore As PizzaStore = New NewYorkPizzaStore()
            Dim chStore As PizzaStore = New ChicagoPizzaStore()

            Dim pizza As Pizza = nyStore.OrderPizza("cheese")
            Console.WriteLine("Ethan ordered a " & pizza.Name + Constants.vbLf)

            pizza = chStore.OrderPizza("cheese")
            Console.WriteLine("Joel ordered a " & pizza.Name + Constants.vbLf)

            pizza = nyStore.OrderPizza("clam")
            Console.WriteLine("Ethan ordered a " & pizza.Name + Constants.vbLf)

            pizza = chStore.OrderPizza("clam")
            Console.WriteLine("Joel ordered a " & pizza.Name + Constants.vbLf)

            pizza = nyStore.OrderPizza("pepperoni")
            Console.WriteLine("Ethan ordered a " & pizza.Name + Constants.vbLf)

            pizza = chStore.OrderPizza("pepperoni")
            Console.WriteLine("Joel ordered a " & pizza.Name + Constants.vbLf)

            pizza = nyStore.OrderPizza("veggie")
            Console.WriteLine("Ethan ordered a " & pizza.Name + Constants.vbLf)

            pizza = chStore.OrderPizza("veggie")
            Console.WriteLine("Joel ordered a " & pizza.Name + Constants.vbLf)

            ' Wait for user
            Console.ReadKey()
        End Sub
	End Class

	#Region "Ingredient Abstract Factory"

	Public Interface IPizzaIngredientFactory
		Function CreateDough() As IDough
		Function CreateSauce() As ISauce
		Function CreateCheese() As ICheese
		Function CreateVeggies() As IVeggies()
		Function CreatePepperoni() As IPepperoni
		Function CreateClam() As IClams
	End Interface

	Public Class NewYorkPizzaIngredientFactory
		Implements IPizzaIngredientFactory
		Public Function CreateDough() As IDough Implements IPizzaIngredientFactory.CreateDough
			Return New ThinCrustDough()
		End Function

		Public Function CreateSauce() As ISauce Implements IPizzaIngredientFactory.CreateSauce
			Return New MarinaraSauce()
		End Function

		Public Function CreateCheese() As ICheese Implements IPizzaIngredientFactory.CreateCheese
			Return New ReggianoCheese()
		End Function

		Public Function CreateVeggies() As IVeggies() Implements IPizzaIngredientFactory.CreateVeggies
			Dim veggies() As IVeggies = { New Garlic(), New Onion(), New Mushroom(), New RedPepper() }
			Return veggies
		End Function

		Public Function CreatePepperoni() As IPepperoni Implements IPizzaIngredientFactory.CreatePepperoni
			Return New SlicedPepperoni()
		End Function

		Public Function CreateClam() As IClams Implements IPizzaIngredientFactory.CreateClam
			Return New FreshClams()
		End Function
	End Class

	Public Class ChicagoPizzaIngredientFactory
		Implements IPizzaIngredientFactory

		Public Function CreateDough() As IDough Implements IPizzaIngredientFactory.CreateDough
			Return New ThickCrustDough()
		End Function

		Public Function CreateSauce() As ISauce Implements IPizzaIngredientFactory.CreateSauce
			Return New PlumTomatoSauce()
		End Function

		Public Function CreateCheese() As ICheese Implements IPizzaIngredientFactory.CreateCheese
			Return New MozzarellaCheese()
		End Function

		Public Function CreateVeggies() As IVeggies() Implements IPizzaIngredientFactory.CreateVeggies
			Dim veggies() As IVeggies = { New BlackOlives(), New Spinach(), New Eggplant() }
			Return veggies
		End Function

		Public Function CreatePepperoni() As IPepperoni Implements IPizzaIngredientFactory.CreatePepperoni
			Return New SlicedPepperoni()
		End Function

		Public Function CreateClam() As IClams Implements IPizzaIngredientFactory.CreateClam
			Return New FrozenClams()
		End Function
	End Class

	#End Region

	#Region "Pizza Stores"

	Public MustInherit Class PizzaStore
		Public Function OrderPizza(ByVal type As String) As Pizza
			Dim pizza As Pizza = CreatePizza(type)
			Console.WriteLine("--- Making a " & pizza.Name & " ---")

			pizza.Prepare()
			pizza.Bake()
			pizza.Cut()
			pizza.Box()

			Return pizza
		End Function

		Public MustOverride Function CreatePizza(ByVal item As String) As Pizza
	End Class

	Public Class NewYorkPizzaStore
		Inherits PizzaStore
		Public Overrides Function CreatePizza(ByVal item As String) As Pizza
			Dim pizza As Pizza = Nothing
			Dim ingredientFactory As IPizzaIngredientFactory = New NewYorkPizzaIngredientFactory()

			Select Case item
				Case "cheese"
					pizza = New CheesePizza(ingredientFactory)
					pizza.Name = "New York Style Cheese Pizza"
				Case "veggie"
					pizza = New VeggiePizza(ingredientFactory)
					pizza.Name = "New York Style Veggie Pizza"
				Case "clam"
					pizza = New ClamPizza(ingredientFactory)
					pizza.Name = "New York Style Clam Pizza"
				Case "pepperoni"
					pizza = New PepperoniPizza(ingredientFactory)
					pizza.Name = "New York Style Pepperoni Pizza"
			End Select
			Return pizza
		End Function
	End Class

	Public Class ChicagoPizzaStore
		Inherits PizzaStore
		' Factory method implementation
		Public Overrides Function CreatePizza(ByVal item As String) As Pizza
			Dim pizza As Pizza = Nothing
			Dim ingredientFactory As IPizzaIngredientFactory = New ChicagoPizzaIngredientFactory()

			Select Case item
				Case "cheese"
					pizza = New CheesePizza(ingredientFactory)
					pizza.Name = "Chicago Style Cheese Pizza"
				Case "veggie"
					pizza = New VeggiePizza(ingredientFactory)
					pizza.Name = "Chicago Style Veggie Pizza"
				Case "clam"
					pizza = New ClamPizza(ingredientFactory)
					pizza.Name = "Chicago Style Clam Pizza"
				Case "pepperoni"
					pizza = New PepperoniPizza(ingredientFactory)
					pizza.Name = "Chicago Style Pepperoni Pizza"
			End Select
			Return pizza
		End Function
	End Class

	#End Region

	#Region "Pizzas"

	Public MustInherit Class Pizza
		Protected dough As IDough
		Protected sauce As ISauce
		Protected veggies() As IVeggies
		Protected cheese As ICheese
		Protected pepperoni As IPepperoni
		Protected clam As IClams

		Private name_Renamed As String

		Public MustOverride Sub Prepare()

		Public Sub Bake()
			Console.WriteLine("Bake for 25 minutes at 350")
		End Sub

		Public Overridable Sub Cut()
			Console.WriteLine("Cutting the pizza into diagonal slices")
		End Sub

		Public Sub Box()
			Console.WriteLine("Place pizza in official Pizzastore box")
		End Sub

		Public Property Name() As String
			Get
				Return name_Renamed
			End Get
			Set(ByVal value As String)
				name_Renamed = value
			End Set
		End Property

		Public Overrides Function ToString() As String
			Dim result As New StringBuilder()
			result.Append("---- " & Me.Name & " ----" & Constants.vbLf)
			If dough IsNot Nothing Then
				result.Append(dough)
				result.Append(Constants.vbLf)
			End If
			If sauce IsNot Nothing Then
				result.Append(sauce)
				result.Append(Constants.vbLf)
			End If
			If cheese IsNot Nothing Then
				result.Append(cheese)
				result.Append(Constants.vbLf)
			End If
			If veggies IsNot Nothing Then
				For i As Integer = 0 To veggies.Length - 1
					result.Append(veggies(i))
					If i < veggies.Length - 1 Then
						result.Append(", ")
					End If
				Next i
				result.Append(Constants.vbLf)
			End If
			If clam IsNot Nothing Then
				result.Append(clam)
				result.Append(Constants.vbLf)
			End If
			If pepperoni IsNot Nothing Then
				result.Append(pepperoni)
				result.Append(Constants.vbLf)
			End If
			Return result.ToString()
		End Function
	End Class

	Public Class ClamPizza
		Inherits Pizza
		Private _ingredientFactory As IPizzaIngredientFactory

		Public Sub New(ByVal ingredientFactory As IPizzaIngredientFactory)
			Me._ingredientFactory = ingredientFactory
		End Sub

		Public Overrides Sub Prepare()
			Console.WriteLine("Preparing " & Me.Name)
			dough = _ingredientFactory.CreateDough()
			sauce = _ingredientFactory.CreateSauce()
			cheese = _ingredientFactory.CreateCheese()
			clam = _ingredientFactory.CreateClam()
		End Sub
	End Class

	Public Class CheesePizza
		Inherits Pizza
		Private _ingredientFactory As IPizzaIngredientFactory

		Public Sub New(ByVal ingredientFactory As IPizzaIngredientFactory)
			Me._ingredientFactory = ingredientFactory
		End Sub

		Public Overrides Sub Prepare()
			Console.WriteLine("Preparing " & Me.Name)
			dough = _ingredientFactory.CreateDough()
			sauce = _ingredientFactory.CreateSauce()
			cheese = _ingredientFactory.CreateCheese()
		End Sub
	End Class

	Public Class PepperoniPizza
		Inherits Pizza
		Private _ingredientFactory As IPizzaIngredientFactory

		Public Sub New(ByVal ingredientFactory As IPizzaIngredientFactory)
			Me._ingredientFactory = ingredientFactory
		End Sub

		Public Overrides Sub Prepare()
			Console.WriteLine("Preparing " & Me.Name)
			dough = _ingredientFactory.CreateDough()
			sauce = _ingredientFactory.CreateSauce()
			cheese = _ingredientFactory.CreateCheese()
			veggies = _ingredientFactory.CreateVeggies()
			pepperoni = _ingredientFactory.CreatePepperoni()
		End Sub
	End Class

	Public Class VeggiePizza
		Inherits Pizza
		Private _ingredientFactory As IPizzaIngredientFactory

		Public Sub New(ByVal ingredientFactory As IPizzaIngredientFactory)
			Me._ingredientFactory = ingredientFactory
		End Sub

		Public Overrides Sub Prepare()
			Console.WriteLine("Preparing " & Me.Name)
			dough = _ingredientFactory.CreateDough()
			sauce = _ingredientFactory.CreateSauce()
			cheese = _ingredientFactory.CreateCheese()
			veggies = _ingredientFactory.CreateVeggies()
		End Sub
	End Class

	#End Region

	#Region "Ingredients"

	Public Class ThinCrustDough
		Implements IDough
		Public Overrides Function ToString() As String Implements IDough.ToString
			Return "Thin Crust Dough"
		End Function
	End Class

	Public Class ThickCrustDough
		Implements IDough
		Public Overrides Function ToString() As String Implements IDough.ToString
			Return "ThickCrust style extra thick crust dough"
		End Function
	End Class

	Public Class Spinach
		Implements IVeggies
		Public Overrides Function ToString() As String Implements IVeggies.ToString
			Return "Spinach"
		End Function
	End Class

	Public Class SlicedPepperoni
		Implements IPepperoni
		Public Overrides Function ToString() As String Implements IPepperoni.ToString
			Return "Sliced Pepperoni"
		End Function
	End Class

	Public Interface ISauce
		Overloads Function ToString() As String
	End Interface
	Public Interface IDough
		Overloads Function ToString() As String
	End Interface
	Public Interface IClams
		Overloads Function ToString() As String
	End Interface
	Public Interface IVeggies
		Overloads Function ToString() As String
	End Interface
	Public Interface ICheese
		Overloads Function ToString() As String
	End Interface

	Public Interface IPepperoni
		Overloads Function ToString() As String
	End Interface

	Public Class Garlic
		Implements IVeggies
		Public Overrides Function ToString() As String Implements IVeggies.ToString
			Return "Garlic"
		End Function
	End Class

	Public Class Onion
		Implements IVeggies
		Public Overrides Function ToString() As String Implements IVeggies.ToString
			Return "Onion"
		End Function
	End Class

	Public Class Mushroom
		Implements IVeggies

		Public Overrides Function ToString() As String Implements IVeggies.ToString
			Return "Mushrooms"
		End Function
	End Class

	Public Class Eggplant
		Implements IVeggies
		Public Overrides Function ToString() As String Implements IVeggies.ToString
			Return "Eggplant"
		End Function
	End Class

	Public Class BlackOlives
		Implements IVeggies
		Public Overrides Function ToString() As String Implements IVeggies.ToString
			Return "Black Olives"
		End Function
	End Class

	Public Class RedPepper
		Implements IVeggies
		Public Overrides Function ToString() As String Implements IVeggies.ToString
			Return "Red Pepper"
		End Function
	End Class

	Public Class PlumTomatoSauce
		Implements ISauce
		Public Overrides Function ToString() As String Implements ISauce.ToString
			Return "Tomato sauce with plum tomatoes"
		End Function
	End Class
	Public Class MarinaraSauce
		Implements ISauce
		Public Overrides Function ToString() As String Implements ISauce.ToString
			Return "Marinara Sauce"
		End Function
	End Class

	Public Class FreshClams
		Implements IClams

		Public Overrides Function ToString() As String Implements IClams.ToString
			Return "Fresh Clams from Long Island Sound"
		End Function
	End Class
	Public Class FrozenClams
		Implements IClams
		Public Overrides Function ToString() As String Implements IClams.ToString
			Return "Frozen Clams from Chesapeake Bay"
		End Function
	End Class

	Public Class ParmesanCheese
		Implements ICheese
		Public Overrides Function ToString() As String Implements ICheese.ToString
			Return "Shredded Parmesan"
		End Function
	End Class

	Public Class MozzarellaCheese
		Implements ICheese
		Public Overrides Function ToString() As String Implements ICheese.ToString
			Return "Shredded Mozzarella"
		End Function
	End Class

	Public Class ReggianoCheese
		Implements ICheese
		Public Overrides Function ToString() As String Implements ICheese.ToString
			Return "Reggiano Cheese"
		End Function
	End Class
	#End Region
End Namespace

File: CSharp/HeadFirstPatterns/DoFactory.HeadFirst/DoFactory.HeadFirst.Factory.Abstract.Pizza/PizzaTestDrive.cs
using System;
using System.Text;

namespace DoFactory.HeadFirst.Factory.Abstract.Pizza
{
    // PizzaTestDrive test application
    class PizzaTestDrive
    {
        static void Main(string[] args)
        {
            PizzaStore nyStore = new NewYorkPizzaStore();
            PizzaStore chStore = new ChicagoPizzaStore();

            Pizza pizza = nyStore.OrderPizza("cheese");
            Console.WriteLine("Ethan ordered a " + pizza.Name + "\n");

            pizza = chStore.OrderPizza("cheese");
            Console.WriteLine("Joel ordered a " + pizza.Name + "\n");

            pizza = nyStore.OrderPizza("clam");
            Console.WriteLine("Ethan ordered a " + pizza.Name + "\n");

            pizza = chStore.OrderPizza("clam");
            Console.WriteLine("Joel ordered a " + pizza.Name + "\n");

            pizza = nyStore.OrderPizza("pepperoni");
            Console.WriteLine("Ethan ordered a " + pizza.Name + "\n");

            pizza = chStore.OrderPizza("pepperoni");
            Console.WriteLine("Joel ordered a " + pizza.Name + "\n");

            pizza = nyStore.OrderPizza("veggie");
            Console.WriteLine("Ethan ordered a " + pizza.Name + "\n");

            pizza = chStore.OrderPizza("veggie");
            Console.WriteLine("Joel ordered a " + pizza.Name + "\n");

            // Wait for user
            Console.ReadKey();
        }
    }

    #region Ingredient Abstract Factory

    public interface IPizzaIngredientFactory
    {
        IDough CreateDough();
        ISauce CreateSauce();
        ICheese CreateCheese();
        IVeggies[] CreateVeggies();
        IPepperoni CreatePepperoni();
        IClams CreateClam();
    }

    public class NewYorkPizzaIngredientFactory : IPizzaIngredientFactory
    {
        public IDough CreateDough()
        {
            return new ThinCrustDough();
        }

        public ISauce CreateSauce()
        {
            return new MarinaraSauce();
        }

        public ICheese CreateCheese()
        {
            return new ReggianoCheese();
        }

        public IVeggies[] CreateVeggies()
        {
            IVeggies[] veggies = { new Garlic(), 
                                   new Onion(), 
                                   new Mushroom(), 
                                   new RedPepper() };
            return veggies;
        }

        public IPepperoni CreatePepperoni()
        {
            return new SlicedPepperoni();
        }

        public IClams CreateClam()
        {
            return new FreshClams();
        }
    }

    public class ChicagoPizzaIngredientFactory : IPizzaIngredientFactory
    {

        public IDough CreateDough()
        {
            return new ThickCrustDough();
        }

        public ISauce CreateSauce()
        {
            return new PlumTomatoSauce();
        }

        public ICheese CreateCheese()
        {
            return new MozzarellaCheese();
        }

        public IVeggies[] CreateVeggies()
        {
            IVeggies[] veggies = { new BlackOlives(), 
                                   new Spinach(), 
                                   new Eggplant() };
            return veggies;
        }

        public IPepperoni CreatePepperoni()
        {
            return new SlicedPepperoni();
        }

        public IClams CreateClam()
        {
            return new FrozenClams();
        }
    }

    #endregion

    #region Pizza Stores

    public abstract class PizzaStore
    {
        public Pizza OrderPizza(string type)
        {
            Pizza pizza = CreatePizza(type);
            Console.WriteLine("--- Making a " + pizza.Name + " ---");

            pizza.Prepare();
            pizza.Bake();
            pizza.Cut();
            pizza.Box();

            return pizza;
        }

        public abstract Pizza CreatePizza(string item);
    }

    public class NewYorkPizzaStore : PizzaStore
    {
        public override Pizza CreatePizza(string item)
        {
            Pizza pizza = null;
            IPizzaIngredientFactory ingredientFactory =
                new NewYorkPizzaIngredientFactory();

            switch (item)
            {
                case "cheese":
                    pizza = new CheesePizza(ingredientFactory);
                    pizza.Name = "New York Style Cheese Pizza";
                    break;
                case "veggie":
                    pizza = new VeggiePizza(ingredientFactory);
                    pizza.Name = "New York Style Veggie Pizza";
                    break;
                case "clam":
                    pizza = new ClamPizza(ingredientFactory);
                    pizza.Name = "New York Style Clam Pizza";
                    break;
                case "pepperoni":
                    pizza = new PepperoniPizza(ingredientFactory);
                    pizza.Name = "New York Style Pepperoni Pizza";
                    break;
            }
            return pizza;
        }
    }

    public class ChicagoPizzaStore : PizzaStore
    {
        // Factory method implementation
        public override Pizza CreatePizza(string item)
        {
            Pizza pizza = null;
            IPizzaIngredientFactory ingredientFactory =
                new ChicagoPizzaIngredientFactory();

            switch (item)
            {
                case "cheese":
                    pizza = new CheesePizza(ingredientFactory);
                    pizza.Name = "Chicago Style Cheese Pizza";
                    break;
                case "veggie":
                    pizza = new VeggiePizza(ingredientFactory);
                    pizza.Name = "Chicago Style Veggie Pizza";
                    break;
                case "clam":
                    pizza = new ClamPizza(ingredientFactory);
                    pizza.Name = "Chicago Style Clam Pizza";
                    break;
                case "pepperoni":
                    pizza = new PepperoniPizza(ingredientFactory);
                    pizza.Name = "Chicago Style Pepperoni Pizza";
                    break;
            }
            return pizza;
        }
    }

    #endregion

    #region Pizzas

    public abstract class Pizza
    {
        protected IDough dough;
        protected ISauce sauce;
        protected IVeggies[] veggies;
        protected ICheese cheese;
        protected IPepperoni pepperoni;
        protected IClams clam;

        private string name;

        public abstract void Prepare();

        public void Bake()
        {
            Console.WriteLine("Bake for 25 minutes at 350");
        }

        public virtual void Cut()
        {
            Console.WriteLine("Cutting the pizza into diagonal slices");
        }

        public void Box()
        {
            Console.WriteLine("Place pizza in official Pizzastore box");
        }

        public string Name
        {
            get { return name; }
            set { name = value; }
        }

        public override string ToString()
        {
            StringBuilder result = new StringBuilder();
            result.Append("---- " + this.Name + " ----\n");
            if (dough != null)
            {
                result.Append(dough);
                result.Append("\n");
            }
            if (sauce != null)
            {
                result.Append(sauce);
                result.Append("\n");
            }
            if (cheese != null)
            {
                result.Append(cheese);
                result.Append("\n");
            }
            if (veggies != null)
            {
                for (int i = 0; i < veggies.Length; i++)
                {
                    result.Append(veggies[i]);
                    if (i < veggies.Length - 1)
                    {
                        result.Append(", ");
                    }
                }
                result.Append("\n");
            }
            if (clam != null)
            {
                result.Append(clam);
                result.Append("\n");
            }
            if (pepperoni != null)
            {
                result.Append(pepperoni);
                result.Append("\n");
            }
            return result.ToString();
        }
    }

    public class ClamPizza : Pizza
    {
        private IPizzaIngredientFactory _ingredientFactory;

        public ClamPizza(IPizzaIngredientFactory ingredientFactory)
        {
            this._ingredientFactory = ingredientFactory;
        }

        public override void Prepare()
        {
            Console.WriteLine("Preparing " + this.Name);
            dough = _ingredientFactory.CreateDough();
            sauce = _ingredientFactory.CreateSauce();
            cheese = _ingredientFactory.CreateCheese();
            clam = _ingredientFactory.CreateClam();
        }
    }

    public class CheesePizza : Pizza
    {
        private IPizzaIngredientFactory _ingredientFactory;

        public CheesePizza(IPizzaIngredientFactory ingredientFactory)
        {
            this._ingredientFactory = ingredientFactory;
        }

        public override void Prepare()
        {
            Console.WriteLine("Preparing " + this.Name);
            dough = _ingredientFactory.CreateDough();
            sauce = _ingredientFactory.CreateSauce();
            cheese = _ingredientFactory.CreateCheese();
        }
    }

    public class PepperoniPizza : Pizza
    {
        private IPizzaIngredientFactory _ingredientFactory;

        public PepperoniPizza(IPizzaIngredientFactory ingredientFactory)
        {
            this._ingredientFactory = ingredientFactory;
        }

        public override void Prepare()
        {
            Console.WriteLine("Preparing " + this.Name);
            dough = _ingredientFactory.CreateDough();
            sauce = _ingredientFactory.CreateSauce();
            cheese = _ingredientFactory.CreateCheese();
            veggies = _ingredientFactory.CreateVeggies();
            pepperoni = _ingredientFactory.CreatePepperoni();
        }
    }

    public class VeggiePizza : Pizza
    {
        private IPizzaIngredientFactory _ingredientFactory;

        public VeggiePizza(IPizzaIngredientFactory ingredientFactory)
        {
            this._ingredientFactory = ingredientFactory;
        }

        public override void Prepare()
        {
            Console.WriteLine("Preparing " + this.Name);
            dough = _ingredientFactory.CreateDough();
            sauce = _ingredientFactory.CreateSauce();
            cheese = _ingredientFactory.CreateCheese();
            veggies = _ingredientFactory.CreateVeggies();
        }
    }

    #endregion

    #region Ingredients

    public class ThinCrustDough : IDough
    {
        public override string ToString()
        {
            return "Thin Crust Dough";
        }
    }

    public class ThickCrustDough : IDough
    {
        public override string ToString()
        {
            return "ThickCrust style extra thick crust dough";
        }
    }

    public class Spinach : IVeggies
    {
        public override string ToString()
        {
            return "Spinach";
        }
    }

    public class SlicedPepperoni : IPepperoni
    {
        public override string ToString()
        {
            return "Sliced Pepperoni";
        }
    }

    public interface ISauce
    {
        string ToString();
    }
    public interface IDough
    {
        string ToString();
    }
    public interface IClams
    {
        string ToString();
    }
    public interface IVeggies
    {
        string ToString();
    }
    public interface ICheese
    {
        string ToString();
    }

    public interface IPepperoni
    {
        string ToString();
    }

    public class Garlic : IVeggies
    {
        public override string ToString()
        {
            return "Garlic";
        }
    }

    public class Onion : IVeggies
    {
        public override string ToString()
        {
            return "Onion";
        }
    }

    public class Mushroom : IVeggies
    {

        public override string ToString()
        {
            return "Mushrooms";
        }
    }

    public class Eggplant : IVeggies
    {
        public override string ToString()
        {
            return "Eggplant";
        }
    }

    public class BlackOlives : IVeggies
    {
        public override string ToString()
        {
            return "Black Olives";
        }
    }

    public class RedPepper : IVeggies
    {
        public override string ToString()
        {
            return "Red Pepper";
        }
    }

    public class PlumTomatoSauce : ISauce
    {
        public override string ToString()
        {
            return "Tomato sauce with plum tomatoes";
        }
    }
    public class MarinaraSauce : ISauce
    {
        public override string ToString()
        {
            return "Marinara Sauce";
        }
    }

    public class FreshClams : IClams
    {

        public override string ToString()
        {
            return "Fresh Clams from Long Island Sound";
        }
    }
    public class FrozenClams : IClams
    {
        public override string ToString()
        {
            return "Frozen Clams from Chesapeake Bay";
        }
    }

    public class ParmesanCheese : ICheese
    {
        public override string ToString()
        {
            return "Shredded Parmesan";
        }
    }

    public class MozzarellaCheese : ICheese
    {
        public override string ToString()
        {
            return "Shredded Mozzarella";
        }
    }

    public class ReggianoCheese : ICheese
    {
        public override string ToString()
        {
            return "Reggiano Cheese";
        }
    }
    #endregion
}
